tablecodec/tablecodec.go

```diff
@@ -388,10 +388,12 @@ func Unflatten(datum types.Datum, ft *types.FieldType, inIndex bool) (types.Datu
 	case mysql.TypeFloat:
 		datum.SetFloat32(float32(datum.GetFloat64()))
 		return datum, nil
+	case mysql.TypeVarchar, mysql.TypeString:
+		datum.SetString(datum.GetString())
```

Comment:
Yes, you are right. After check the MySQL `Data Types Reference`, I realized that the `Collate` and `Charset` is the main difference between `CHAR` and `BINARY`. 

But it seems like that `Unflatten` doesn't care the `Collate` defined in `FieldType`. That result in functions which need take different strategy to handle binary string and nonbinary string are difficult to distinguish them. I think we can set `Datum.k` and `Datum.collation` in `Unflatten` based on `Collate` and `Charset` in `FieldType`.
This change let functions easy to distinguish binary string and nonbinary string, while still let functions to judge what kind of datum based on charset and collate.

For bug mentioned in this PR, I finally find that MySQL treat all string as binary string when convert string to bit. So I revert previous commit and fix this bug in a new commit.